lib/client/pgservicefile/servicefile.go

```diff
@@ -0,0 +1,273 @@
+// service file.
+func Delete(name string) error {
+	serviceFile, err := Load("")
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	err = serviceFile.Delete(name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// ServiceFile represents Postgres connection service file.
+//
+// https://www.postgresql.org/docs/13/libpq-pgservice.html
+type ServiceFile struct {
+	// iniFile is the underlying ini file.
+	iniFile *ini.File
+	// path is the service file path.
+	path string
+}
+
+// Load loads Postgres connection service file from the provided path or the
+// default location if it's not provided.
+func Load(path string) (*ServiceFile, error) {
+	// If the file path wasn't provided, use the default location which
+	// is .pg_service.conf file in the user's home directory.
+	if path == "" {
+		// TODO(r0mant): Check PGSERVICEFILE and PGSYSCONFDIR env vars as well.
+		user, err := user.Current()
+		if err != nil {
+			return nil, trace.ConvertSystemError(err)
+		}
+		path = filepath.Join(user.HomeDir, pgServiceFile)
+	}
+	// Loose load will ignore file not found error.
+	iniFile, err := ini.LooseLoad(path)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return &ServiceFile{
+		iniFile: iniFile,
+		path:    path,
+	}, nil
+}
+
+// Add adds the provided connection profile to the service file and saves it.
+//
+// The profile goes into a separate section with the name equal to the
+// name of the database that user is logged into and looks like this:
+//
+//   [postgres]
+//   host=proxy.example.com
+//   port=3080
+//   sslmode=verify-full
+//   sslrootcert=/home/user/.tsh/keys/proxy.example.com/certs.pem
+//   sslcert=/home/user/.tsh/keys/proxy.example.com/alice-db/root/aurora-x509.pem
+//   sslkey=/home/user/.tsh/keys/proxy.example.com/user
+//
+// With the profile like this, a user can refer to it using "service" psql
+// parameter:
+//
+//   $ psql "service=postgres <other parameters>"
+func (s *ServiceFile) Add(profile ConnectProfile) error {
+	section := s.iniFile.Section(profile.Name)
+	if section != nil {
+		s.iniFile.DeleteSection(profile.Name)
+	}
+	section, err := s.iniFile.NewSection(profile.Name)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	section.NewKey("host", profile.Host)
+	section.NewKey("port", strconv.Itoa(profile.Port))
+	if profile.User != "" {
+		section.NewKey("user", profile.User)
+	}
+	if profile.Database != "" {
+		section.NewKey("dbname", profile.Database)
+	}
+	section.NewKey("sslmode", profile.SSLMode)
+	section.NewKey("sslrootcert", profile.SSLRootCert)
+	section.NewKey("sslcert", profile.SSLCert)
+	section.NewKey("sslkey", profile.SSLKey)
+	ini.PrettyFormat = false // Pretty format breaks psql.
+	return s.iniFile.SaveTo(s.path)
+}
+
+// Env returns the specified connection profile information as a set of
+// environment variables recognized by Postgres clients.
+func (s *ServiceFile) Env(name string) (map[string]string, error) {
+	section, err := s.iniFile.GetSection(name)
```

Comment:
I love typing :) Moreover, I need a map rather than a struct and also to verify that all keys that need to be present are present to avoid broken service configurations so some amount of laborious parsing is unavoidable.